### PR TITLE
Fix weight tying in TF-ESM

### PR DESCRIPTION
**Pull request by Rocketknight1**
TF ESM cloned weights instead of tying, which worked when loading from PT but broke when loading from safetensors. This resolves the issue by correctly tying weights when this is enabled in the config. Fixes an ongoing CI error raised by @ydshieh

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Also cc @gante in case he hates how I handled weight tying here, I don't want to break TF convention too much!

**Comment by gante** [comment]
@Rocketknight1 I'm cool with this :D

**Review by sgugger — APPROVED** [review]
LGTM but @amyeroberts might have more insight in the TF subtleties of this code.

File: tests/models/esm/test_modeling_tf_esm.py [hunk 1]
```diff
@@ -262,6 +262,24 @@ def test_resize_token_embeddings(self):
     def test_save_load_after_resize_token_embeddings(self):
         pass
 
+    def test_model_common_attributes(self):
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+
+        for model_class in self.all_model_classes:
+            model = model_class(config)
```
Comment (amyeroberts) [inline tests/models/esm/test_modeling_tf_esm.py:269]: The modification of the modeling code is controlled by `self.config.tie_word_embeddings`, but all the models here are using the same config. Is the toggling of tying or not tying the weights tested elsewhere?

I'm quite likely missing something just judging on the diff here though

File: src/transformers/models/esm/modeling_tf_esm.py [hunk 2]
```diff
@@ -1102,6 +1103,11 @@ def __init__(self, config):
 
         self.esm = TFEsmMainLayer(config, add_pooling_layer=False, name="esm")
         self.lm_head = TFEsmLMHead(config, name="lm_head")
+        if config.tie_word_embeddings:
+            # Ensure word embeddings are built so that we actually have something to tie
+            with tf.name_scope(os.path.join(self._name_scope(), "esm", "embeddings", "word_embeddings")):
```
Comment (amyeroberts) [inline src/transformers/models/esm/modeling_tf_esm.py:1108]: Cries in TensorFlow 😭

**Review by amyeroberts — APPROVED** [review]
All looks good to me! Thanks for adding this :) 

I just have a small question about the tests.

File: src/transformers/models/esm/modeling_tf_esm.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/models/esm/modeling_tf_esm.py:1108]: I know, but it's the only option when we want to create a weight for a distant sublayer! TF only 'walks' the name scope hierarchy via the call stack when weights are built during `call()`; every other time you have to explicitly enter the `tf.name_scope()` you want.

File: tests/models/esm/test_modeling_tf_esm.py (hunk 1 above)
Comment (Rocketknight1) [reply tests/models/esm/test_modeling_tf_esm.py:269]: The reason for overriding this test is that the common test expects `model.get_output_embedding()` to return a `tf.keras.layers.Layer`, but in this case I'm using a simple shared matrix created with `add_weight`, so I had to tweak the test a little. I'm actually not testing the effect of that parameter anywhere, but maybe I should, unless it's already tested somewhere in the common tests!

**Pull request merged**